### ISSUE TITLE
CPU load is not displayed

### ISSUE THREAD

**Issue by Kefisto** [hetrixtools/agent-windows#1]
![{FA2B0411-53F2-45A3-8EE2-817E797F3073}](https://github.com/user-attachments/assets/19d5e4a1-f2ae-4f20-bb63-a9fea312b664)

**Comment by flash608** [comment]
Same here on my Windows installation (Intel NUC). No usage, nor temp of CPU.

**Comment by hetrixtools** [comment]
Regarding CPU usage, this is a known issue for non-English Windows systems.

Regarding the CPU temperature, it hasn't been implemented yet in our Windows agent.

**Comment by flash608** [comment]
> Regarding CPU usage, this is a known issue for non-English Windows systems.
> 
> Regarding the CPU temperature, it hasn't been implemented yet in our Windows agent.

Oh, thank you for such a quick response :)
I'm glad that you're aware of it already. Hope it will be working some time.

Best regards
Przemek

**Comment by Florian183** [comment]
Hello. First, I want to thank you for your product. I've now used it for a few days and I'm quite impressed.

Second, I also encountered this issue on a system with a German Windows OS.
After some research I found a possible solution: in order to get the localized names of the performance counter, you can add this function to the hetrixtools_agent.ps1 script:

```
function Get-LocalPerfCounterName {

   param (
      [Parameter(Mandatory=$true)]
      $Name
    )

   $key009 = 'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Perflib\009'
   $counters009 = (Get-ItemProperty -Path $key009 -Name Counter).Counter.tolower()
   $Index = $counters009.IndexOf($name.tolower())
   if ($index -eq -1) {
      $null   # not found
   }
   else {
      $keylocal = 'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Perflib\CurrentLanguage'
      $counterslocal = (Get-ItemProperty -Path $keylocal -Name Counter).Counter
      $counterslocal[$index]
   }
}
```

This function will look into the registry and get the ID of the performance counter specified by its English name. It will then determine the localized name using the previously determined ID.

You can then replace the English names in the script with the localized names:
```
[...]
# Initialise values
$total_cpuUsage = 0
$total_diskTime = 0

# determine localized names of CPU performance counters
$processor = Get-LocalPerfCounterName "Processor"
$processorTime = Get-LocalPerfCounterName "% Processor Time"

# Collect data loop
for ($X = 1; $X -le $RunTimes; $X++) {
    # Start both commands as jobs
    $cpuJob = Start-Job -ScriptBlock { 
        $cpuCounter = Get-Counter "\$using:processor(_Total)\$using:processorTime" -SampleInterval $using:CollectEveryXSeconds
        return $cpuCounter.CounterSamples.CookedValue
    }
[...]
```

Credit goes mainly to this website (sorry, it's in German):
https://www.msxfaq.de/powershell/psperfcounter.htm

**Comment by flash608** [comment]
> Hello. First, I want to thank you for your product. I've now used it for a few days and I'm quite impressed.
> 
> Second, I also encountered this issue on a system with a German Windows OS. After some research I found a possible solution: in order to get the localized names of the performance counter, you can add this function to the hetrixtools_agent.ps1 script:
> 
> ```
> function Get-LocalPerfCounterName {
> 
>    param (
>       [Parameter(Mandatory=$true)]
>       $Name
>     )
> 
>    $key009 = 'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Perflib\009'
>    $counters009 = (Get-ItemProperty -Path $key009 -Name Counter).Counter.tolower()
>    $Index = $counters009.IndexOf($name.tolower())
>    if ($index -eq -1) {
>       $null   # not found
>    }
>    else {
>       $keylocal = 'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Perflib\CurrentLanguage'
>       $counterslocal = (Get-ItemProperty -Path $keylocal -Name Counter).Counter
>       $counterslocal[$index]
>    }
> }
> ```
> 
> This function will look into the registry and get the ID of the performance counter specified by its English name. It will then determine the localized name using the previously determined ID.
> 
> You can then replace the English names in the script with the localized names:
> 
> ```
> [...]
> # Initialise values
> $total_cpuUsage = 0
> $total_diskTime = 0
> 
> # determine localized names of CPU performance counters
> $processor = Get-LocalPerfCounterName "Processor"
> $processorTime = Get-LocalPerfCounterName "% Processor Time"
> 
> # Collect data loop
> for ($X = 1; $X -le $RunTimes; $X++) {
>     # Start both commands as jobs
>     $cpuJob = Start-Job -ScriptBlock { 
>         $cpuCounter = Get-Counter "\$using:processor(_Total)\$using:processorTime" -SampleInterval $using:CollectEveryXSeconds
>         return $cpuCounter.CounterSamples.CookedValue
>     }
> [...]
> ```
> 
> Credit goes mainly to this website (sorry, it's in German): https://www.msxfaq.de/powershell/psperfcounter.htm

Well, it started showing something for CPU usage, but in my case the readings are too low. My CPU is currently used in range of 25-45% whereas the graph shows 2,42 - 4,92%... like it's 10x too low.

Still I got no IOWait measure/graph.

Good job either way :)

EDIT:
Since it was around 10x too low, I've multiplied it by 10 :) but I guess it's not a good/best solution. Sorry I'm have no coding/programming skills.


```
[...]
# Calculate CPU Usage
$cpuUsage = [math]::Round($total_cpuUsage / $X * 10, 2)

# Get the disk I/O wait time
$diskTime = [math]::Round($total_diskTime / $X * 10, 2)
[...]
```

**Comment by flash608** [comment]
on the right side of CPU graph, you can see before and after 10x multiply.
![Image](https://github.com/user-attachments/assets/6f2e4fe7-f288-4443-803d-237ec3380f6a)


after longer time
![Image](https://github.com/user-attachments/assets/460aa7e7-d1b9-45ba-9529-46ef42c125db)

**Comment by Florian183** [comment]
@flash608 
You are right. The values do not reflect what the Task Manager shows. Apparently, Microsoft changed the values shown in Task Manager after Windows 7. (source: https://learn.microsoft.com/en-us/troubleshoot/windows-client/performance/cpu-usage-exceeds-100).

I changed the script so that instead of using `\Processor(_Total)\% Processor Time` I now use `\Processor Information(_Total)\% Processor Utility`.

I also had to change the function `Get-LocalPerfCounterName` a little bit because the length of the array in the English language may not always be the same as in the CurrentLanguage. This means I have to first get the correct ID for the name and then look for the ID in the array with the German texts and get the correct name from there.

I also applied the function to the names for the Disk counters accordingly, so these values show up in the HetrixTools status, too.

The function looks like this now:

```
function Get-LocalPerfCounterName {
   param (
      [Parameter(Mandatory=$true)]
      $Name
    )

   $key009 = 'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Perflib\009'
   $counters009 = (Get-ItemProperty -Path $key009 -Name Counter).Counter.tolower()
   $index = $counters009.IndexOf($name.tolower())
   
   if ($index -eq -1) {
      $null   # not found
   }
   else {
      $idOfName = $counters009[$index - 1]
      $keylocal = 'Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Perflib\CurrentLanguage'
      $counterslocal = (Get-ItemProperty -Path $keylocal -Name Counter).Counter
      $indexOfLocalId = $counterslocal.IndexOf($idOfName)
      $counterslocal[$indexOfLocalId + 1]
   }
}
```

The block for CPU and disk counters looks like this now:
```
[...]
# determine localized names of CPU performance counters
$processorInformationName = Get-LocalPerfCounterName "Processor Information"
$processorUtilityName = Get-LocalPerfCounterName "% Processor Utility"

# determine localized names of Disk performance counters
$physicalDiskName = Get-LocalPerfCounterName "PhysicalDisk"
$diskTimeName = Get-LocalPerfCounterName "% Disk Time"

# Collect data loop
for ($X = 1; $X -le $RunTimes; $X++) {
    # Start both commands as jobs
    $cpuJob = Start-Job -ScriptBlock { 
        $cpuCounter = Get-Counter "\$using:processorInformationName(_Total)\$using:processorUtilityName" -SampleInterval $using:CollectEveryXSeconds
        return $cpuCounter.CounterSamples.CookedValue
    }

    $diskJob = Start-Job -ScriptBlock { 
        $diskCounter = Get-Counter "\$using:physicalDiskName(_Total)\$using:diskTimeName" -SampleInterval $using:CollectEveryXSeconds
        return $diskCounter.CounterSamples.CookedValue
    }

    # Wait for both jobs to complete
    $cpuJob | Wait-Job
    $diskJob | Wait-Job
[...]
```

**Comment by flash608** [comment]
> so that instead of using

That works correctly :) great job

CPU Usage in Hetrix is reported the same as the Task Manager is showing.

![Image](https://github.com/user-attachments/assets/5c366388-848e-4501-978c-dc33b5f40f84)

On the right side of the graph is a little 52% and 99% bump/peak - I've set in Power Plan max. CPU state to 50 and 99% - it does work that way to test it. Below is a screen from my Task Manager which shows 52% usage.

![Image](https://github.com/user-attachments/assets/aa30bc2e-a6da-4e53-a804-954b13a8d2b6)

I have two topics more.
1. HwInfo is showing 100% CPU usage whereas Task Manager show 52% - OK, I get it, HwInfo is getting the CPU Usage from different source, would be great if both would show the same value

**[EDIT] In HwInfo it's "Core Utility" or "Total CPU Utility" - it's the same value as the CPU usage from Task Manager**

2. If my CPU Usage is high e.g. 100% using CPU-z then the task from Task Scheduler is not working at all/not being refreshed/not executed, so I do not get any data in Hetrix. I had to set CPU-z process to as low as possible in Task Manager, and run the Hetrix ps1 code from PowerShell ISE directly to be able to send the data to Hetrix.
In other words, tasks from Task Scheduler are not executed when CPU usage it too high.

**Comment by hetrixtools** [comment]
We'll look to implement the fix proposed by @Florian183 in a future agent release. Thank you, everyone, for the feedback provided in this thread.

**Comment by Florian183** [comment]
> 2. If my CPU Usage is high e.g. 100% using CPU-z then the task from Task Scheduler is not working at all/not being refreshed/not executed, so I do not get any data in Hetrix. I had to set CPU-z process to as low as possible in Task Manager, and run the Hetrix ps1 code from PowerShell ISE directly to be able to send the data to Hetrix.
>    In other words, tasks from Task Scheduler are not executed when CPU usage it too high.

(I suggest moving that part into a separate issue.)

That behavior is normal for Task Scheduler tasks. By default, tasks in Task Scheduler are created with Priority 7, which means that they will not run normally when the CPU is completely busy with other things.

Source: https://learn.microsoft.com/en-us/windows/win32/taskschd/tasksettings-priority

It's possible to change the priority of Task Scheduler tasks. You can either export the task to XML, change the priority in the XML and import the task again, or you can use Powershell to do that:

Source: https://stackoverflow.com/questions/47197821/how-to-change-default-scheduled-task-process-priority-in-windows

Essentially, in Powershell you can do something like this:
```
$taskName = "HetrixTools Server Monitoring Agent"
$task = Get-ScheduledTask -TaskName $taskName
$settings = $task.Settings
$settings.Priority = 4
Set-ScheduledTask -TaskName $taskName -Settings $settings
```

**Disclaimer**: Since this change may affect the performance of your system I suggest testing this thoroughly before applying this to a production server.

**Comment by hetrixtools** [comment]
When we were testing this fix on several different versions we've found that it did not fully work on Spanish Windows versions.

However, we were able to find a much simpler fix for this issue, just replacing the names with the ids in the agent query (ids being independent of system language).

Implemented on v2.0.6:
https://github.com/hetrixtools/agent-windows/commit/be96f7810dfc682cd43f76bc8bd194c5860afd8d

Thank you all for contributing to this issue.

**Comment by flash608** [comment]
Hi, thank you for the update.

In my case the CPU reading is now showing up but it's like 10x less than it should be and I don't get IOWait reading at all.

On the picture:

area 1 - it was after implementing Florian183 suggestion + multiply CPU and IOWait *10 times

area 2 - 2.0.6 code

area 3 - 2.0.6 code with *10 times CPU & IOWait

![IMG-20250422-WA0000.jpg](https://github.com/user-attachments/assets/3b86ad45-3ec7-4152-884f-3857448482c8)

**Comment by hetrixtools** [comment]
Can you try the latest version, `v2.0.7`, and see if this happens?

Also, this issue seems to be different from the one in this thread. If the problem persists, please open a new issue.

**Comment by flash608** [comment]
![Image](https://github.com/user-attachments/assets/d92dab1c-d76a-4475-a739-d28b9ff56c35)
It does work for CPU/IOWait, as in the picture but still the reading are around 10x too low.
